repository/httpRemoteRepository.go

```diff
@@ -0,0 +1,318 @@
+	cacheDirectory := filepath.Join(r.endpointCacheDir, doguVersion.Name.String(), doguVersion.Version.Raw)
+	return r.receiveDoguFromRemoteOrCache(requestUrl, cacheDirectory)
+}
+
+func (r *httpRemote) receiveDoguFromRemoteOrCache(requestUrl string, dirname string) (*core.Dogu, error) {
+	var remoteDogu, err = r.readCachedDogu(dirname)
+	if err != nil {
+		err = r.retrier.Run(func() error {
+			requestErr := r.requestWithoutCredentialsFirst(requestUrl, &remoteDogu)
+			if requestErr != nil {
+				return r.request(requestUrl, &remoteDogu, true)
+			}
+			return requestErr
+		})
+
+		if errors.Is(err, errNotFound) {
+			return nil, dogu.ErrDescriptorNotFound
+		}
+
+		err = r.writeDoguToCache(remoteDogu, dirname)
+		if err != nil {
+			return &core.Dogu{}, fmt.Errorf("failed to write dogu to cache: %w", err)
+		}
+	}
+
+	return remoteDogu, nil
+}
+
+func (r *httpRemote) readCachedDogu(dirname string) (*core.Dogu, error) {
+	if r.useCache {
+		cacheFile := filepath.Join(dirname, "content.json")
+		doguFromFile, _, err := core.ReadDoguFromFile(cacheFile)
+		if err != nil {
+			return &core.Dogu{}, fmt.Errorf("failed to read from remote registry and cache: %w", err)
+		}
+		return doguFromFile, nil
+	}
+	return &core.Dogu{}, fmt.Errorf("useCache is not activated")
+}
+
+func (r *httpRemote) writeDoguToCache(dogu *core.Dogu, dirname string) error {
+	err := os.MkdirAll(dirname, os.ModePerm)
+	if nil != err {
+		return fmt.Errorf("failed to create cache directory %s: %w", dirname, err)
+	}
+
+	cacheFile := filepath.Join(dirname, "content.json")
+	err = core.WriteDoguToFile(cacheFile, dogu)
+
+	if nil != err {
+		removeErr := os.Remove(cacheFile)
+		if removeErr != nil {
+			core.GetLogger().Warningf("failed to remove cache file %s", cacheFile)
+		}
+		return fmt.Errorf("failed to write cache %s: %w", cacheFile, err)
+	}
+
+	return nil
+}
+
+func (r *httpRemote) requestWithoutCredentialsFirst(requestURL string, responseType interface{}) error {
+	core.GetLogger().Debug("Access \"" + requestURL + "\" anonymous...")
+	err := r.request(requestURL, responseType, false)
+	if err != nil {
+		core.GetLogger().Debug("Anonymous access to \"" + requestURL + "\" failed. Using credentials...")
+		err = r.request(requestURL, responseType, true)
+		if err != nil {
+			core.GetLogger().Debug("Access to \"" + requestURL + "\" with credentials failed...")
+		} else {
+			core.GetLogger().Debug("Access to \"" + requestURL + "\" with credentials was successfull...")
+		}
+	}
+	return err
+}
+
+func (r *httpRemote) request(requestURL string, responseType interface{}, useCredentials bool) error {
+	core.GetLogger().Debugf("fetch json from remote %s", requestURL)
+
+	request, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		return fmt.Errorf("failed to prepare request: %w", err)
+	}
+
+	if useCredentials && r.credentials != nil {
+		request.SetBasicAuth(r.credentials.Username, r.credentials.Password)
+	}
+
+	resp, err := r.client.Do(request)
```

Comment:
🚫 **[checkstyle]** <bodyclose> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>response body must be closed
<!-- __reviewdog__:ChAzZjUxMzRiNWM5ZDNlNzVhEgpjaGVja3N0eWxl -->